weed/util/fullpath.go

```diff
@@ -6,15 +6,19 @@ import (
 	"strings"
 )
 
+// FullPath will keep the literal passed in.
+// file literal: /path/to/file
+// directory literal: /path/
```

Comment:
The previous assumption was to have `/path/to/file` format always. The fullpath will not end with "/" unless it is root.

Reply:
Scenarios like below need path literal to represent a path.
https://github.com/chrislusf/seaweedfs/blob/b40d252761e92387b281b727d490088ad0bfdab5/weed/server/filer_server_handlers_write_autochunk.go#L147-L160

Reply:
Current logic works well. What it needs now?

Reply:
For seaweed, it's better to have a **uniform** way (using this fullpath "class", which is more clean and readable) rather than string operation (which is more error-prone) to handle path (**resolve special dir like "." and ".."; replace invalid utf-8 characters with "?"**).

And in fullpath, we need to **distinguish directories from files**. For example, user upload file to dir instead of file by http request. (upload xxx.txt to /x/y/). Supporting dir literal will not save files uploaded to dir into file (save to /x/y mistakenly).

Reply:
The current uniform way is to use `/parent/dir/entry_name`, where entry_name can be a file or a directory.

Supporting both `/x/y` and /x/y/` will be confusing.

Reply:
Almost all handlers use fullpath to resolve path. So fullpath is a "common place" to add path resolve.

> The path resolving, cleaning or distinguishing problem can be solved **before the path "entering" the system.**

If we choose other way, there is no such "common place".
For http server, we can add code like `r.URL.Path = clearPath(r.URL.Path)` in `filer_server_handlers.go` after line 16.
But for grpc server, there is no such "common place". We have to clear path in every handle respectively.
https://github.com/chrislusf/seaweedfs/blob/f70c1e449b9f80545ff7e6c1c88f2c2e27e4d0e8/weed/server/filer_server_handlers.go#L16-L25

Reply:
The change is trying too many things and no clear assumption. This will add more confusion.

Reply:
> The current uniform way is to use `/parent/dir/entry_name`, where entry_name can be a file or a directory.
> 
> Supporting both `/x/y` and `/x/y/` will be confusing. There are no sure way to tell `/x/y` is a directory or not. Having a `/x/y/` will only add more confusion.
> 
> The path resolving, cleaning or distinguishing problem can be solved before the path "entering" the system.

OK! Follow this uniform and use `NewFullPath` manually to clean path before use the path.